Example for admin metadata implementation,

dantleech opened the pull request:
See #18 
## Descriptors

A description is just an key / value store and passed to any number of
supprorting "enrichers" (e.g. SonataEnricher, SyliusEnricher) which add any
information.

In order to ensure interoperability a `Descriptor` class is given which
provides constant values for use as descriptor keys, e.g. `LINK_EDIT_HTML`,
`TYPE_ALIAS`, `TITLE`.

Use is as follows:

``` php
$factory = new DescriptionFactory([
    new SyliusEnricher(...),  // add links for Sylius
    new IconEnricher()        // add "icon" URLs (as an example)
]); 

$resource = $registry->get('/pages/foobar/my-article'); // get a resource
$description = $factory->getPayloadDescriptionFor($resource);

echo $description->get(Descriptor::TITLE); // "My Article"
echo $description->get(Descriptor::TYPE); // AcmeBundle\Document\Article
echo $description->has(Descriptor::LINK_HTML_EDIT); // true
echo $description->get(Descriptor::LINK_HTML_EDIT); // /path/to/edit/page/123
```
## Enhancers

Enhancers included in this PR:
- SonataAdminEnhancer: Adds CRUD URLs for resources.
- SyliusEnhancer: Adds CRUD URLs for resources.

dantleech committed:
Admin metadata

ElectricMaxxx commented on Metadata/PayloadMetadata.php:
```diff
@@ -0,0 +1,68 @@
+     */
+    public function get($key)
+    {
+        if (!isset($this->supported[$key])) {
```
Where would you fill that array?

dantleech replied:
That was just junk from the WIP.

ElectricMaxxx commented on Metadata/PayloadMetadata.php:
```diff
@@ -0,0 +1,68 @@
+    public function __construct($providerName, array $metadata)
+    {
+        $this->providerName = $providerName;
+        $this->metadata = $metadata;
```
`$metadata` is a kind of a configuration from a bundle view? Where would you divide the paylod-type depending stuff. Maybe provide REST api links for StaticContent, but not for MySpecialSecretContent (phpcr-odm view)?

ElectricMaxxx replied:
ahh the metadata factories are responsible for that, but i would still miss a kind of a configuration/parameter input to enable/restrict some types, or? Sure on Factory Level.

dantleech replied:
Have now refactored completely, each "enricher" is now responsible for declaring which resources it supports.

dantleech committed:
Refactored

dantleech commented:
Have now refactored (and updated the description of this PR).

I am tempted to leave the implementations of the "enrichers" out of this component, and put them in separate packages.

@WouterJ would be great if you could have a look and let me know if you agree with this direction.

dantleech commented on Description/DescriptionEnricherInterface.php:
```diff
@@ -0,0 +1,36 @@
+
+use Symfony\Cmf\Component\Resource\Repository\Resource\CmfResource;
+
+interface DescriptionEnricherInterface
```
Consider renaming this with the more apt term (imo) "augmenter"

dantleech replied:
Or use Enhancer as with the RoutingBundle - I think the semantic is the same /cc @dbu

dantleech committed:
Removed 5.4 support

wouterj commented on Description/DescriptionFactory.php:
```diff
@@ -0,0 +1,49 @@
+    public function getPayloadDescriptionFor(CmfResource $resource)
+    {
+        $type = $resource->getPayloadType();
+        $payload = $resource->getPayload();
```
<s>we have to check if payload is an object. At the moment, it can be anything: https://github.com/symfony-cmf/resource/blob/master/Repository/Resource/CmfResource.php#L40-L48 (I even believe we use it as string in the ResourceBundle atm)</s>

That was a wrong comment. However, I don't see why we have to know the payload? Removing this and injecting the resource in the enricher would allow us to use any puli resource.

dantleech replied:
In that case we would need to inject tge Resource into the Description, making it a Resource Description and not a Payload Description (what I see it as now) but, what you suggest probably makes more sense.

dantleech replied:
Updated, the `Description` now accepts a `PuliResource` and we pass only the `Description` to the `enhance` (renamed from `enrich`) method of the `Enhancer`.

dantleech committed:
Allow PuliResources

dantleech commented on Description/Description.php:
```diff
@@ -0,0 +1,89 @@
+    /**
+     * Return the descriptors value for the given key.
+     *
+     * The key should be one of the constants defined in this class.
```
Not "this" class.

dantleech commented on Description/Description.php:
```diff
@@ -0,0 +1,89 @@
+     * - It is possible to overwrite existing keys.
+     *
+     * - To help ensure interoperability, where possible, the key should be the
+     *   value of one of the appropriate constants defined in the MetadescriptorsKey
```
in the `Descriptor` class

dantleech commented on Description/DescriptionFactory.php:
```diff
@@ -0,0 +1,50 @@
+class DescriptionFactory
+{
+    /**
+     * @var Enhancer[]
```
missing use

dantleech commented on Description/Descriptor.php:
```diff
@@ -0,0 +1,52 @@
+    const PAYLOAD_TITLE = 'title';
+
+    /**
+     * Keys to be used to store link values for HTML.
```
Refer to these as `descriptors`, not "keys"

dantleech commented:
Updated.

wouterj commented on Description/Description.php:
```diff
@@ -0,0 +1,88 @@
+    public function set($key, $value)
+    {
+        $this->descriptors[$key] = $value;
+    }
```
What about also implementing the other methods mentioned in http://symfony.com/doc/current/contributing/code/conventions.html#method-names ?

dantleech replied:
Yeah we are missing `has()` and possibly `all()`.

dbu replied:
something to set several descriptors in one call, by providing a hashmap that is merged with $this->descriptors?

dantleech replied:
I guess that would be `merge()` then.

dbu replied:
or setDescriptor($key, $value) and setDescriptors($descriptors)

dantleech replied:
But would `setDescriptors` replace all the descriptors or merge them? I would assume without prior knowledge that it would replace them.

dbu replied:
ah, true. maybe addDescriptors? mergeDescriptors feels strange. that adding something that already exists overwrites the existing value seems no big surprise to me, and we can state it explicitly in the phpdoc

dantleech replied:
I prefer merge, as that correlates with `array_merge` which is what you are actually doing. However, I think the `ParameterBag` uses `add()` with the same semantic.

I guess suffixing `Descriptor` makes sense.

dbu replied:
your call :-)

wouterj commented on Description/DescriptionFactory.php:
```diff
@@ -0,0 +1,50 @@
+                continue;
+            }
+
+            $enhancer->enhance($description);
```
I'm not a big fan of these by reference edits. What about doing `$description = $enhancer->enhance($description);`? That would be more flexible.

dantleech replied:
`$description` is an object, even if you did return it it would still be modified by reference - unless we made it immutable. But I don't really see the benefit.

wouterj replied:
We had the same thing in RoutingAuto, which we later deprecated because it turned out that some people needed to not modify by reference: https://github.com/symfony-cmf/routing-auto/pull/53

dantleech replied:
Yes, but in that case it was the `Entity` which came from "outside". In this case, we create the `Description`, nobody is relying on its state before this factory produces it.

dantleech replied:
And returning or not returning, the result will always be the same and the `Description` will always be the same instance.

wouterj commented:
Although I understand the reasons behind the `Descriptor` constants, I'm afraid blocks the current limits the amount of options that enhancers can edit.

Also, why not add these constants to the `Description` class?

dantleech commented:
> Although I understand the reasons behind the Descriptor constants, I'm afraid blocks the current limits the amount of options that enhancers can edit.

My main concern is to try and avoid a situation where one implementation uses a different key/descriptor than another implementation for the same thing.

But I don't really see any alternative to defining a canonical set of "descriptors". How else could we "solve" this problem?

> Also, why not add these constants to the Description class?

It just seemed a _little_ bit more semantic:

``` php
$description->set(Descriptor::TITLE, 'Hello');
$description->set(Descriptor::LINK_EDIT_HTML, 'http://foo/bar.html');
```

vs.

``` php
$description->set(Description::TITLE, 'Hello');
$description->set(Description::LINK_EDIT_HTML, 'http://foo/bar.html');
```

and it separated the responsiblitiy of storing these values into a separate
class which just seems tidier.

Don't feel too strongly about it however.

dbu commented on Description/Description.php:
```diff
@@ -0,0 +1,88 @@
+class Description
+{
+    /**
+     * @var array
```
what are the values in this array? strings, or something more complex? a list of label => url?

dantleech replied:
I think we can limit them to be scalar values. If there is a use case for array values we could allow that later.

dantleech replied:
Limited to scalar values only

dantleech replied:
We should allow arrays also, f.e. to indicate the valid child payload types.

dantleech replied:
Done.

dbu commented on Description/Description.php:
```diff
@@ -0,0 +1,88 @@
+     *
+     * @param string $key
+     *
+     * @return mixed
```
can we be more specific what this is? how can the consumer work with the result?

dantleech replied:
well, the consumer can use the result in any way they please - for example the result could be the name of an `.png` file to use as an icon, or it could be the URL, or a vendor-assigned human title for the resource class (e.g. "Page"). Its pretty generic.

dbu replied:
so you would split parts of the information into separate descriptions? like edit.html.icon and edit.html.url?

dantleech replied:
Exactly, see the `Descriptor` class.

Then, when something needs a certain piece of information, it calls `$description->has(Descriptor::LINK_HTML_EDIT)`.

dbu commented on Description/DescriptionFactory.php:
```diff
@@ -0,0 +1,50 @@
+     * Return a description of the given (CMF) Resource.
+     *
+     * @param PuliResource $resource
+     */
```
missing `@return`

dantleech commented:
As for implementations, I don't really want to include both Sonata and Sylius as dev-dependencies. I experimented by creating [this repository](https://github.com/symfony-cmf/resource-descriptor-enhancer-sonata) for Sonata integration.

I imagine Travis running the tests for these packages as part of this packages build. Open to suggestions ..

dantleech commented on Description/Descriptor.php:
```diff
@@ -0,0 +1,52 @@
+    const LINK_CREATE_HTML = 'link.create.html';
+    const LINK_UPDATE_HTML = 'link.update.html';
+    const LINK_REMOVE_HTML = 'link.remove.html';
+    const LINK_SHOW_HTML = 'link.show.html';
```
missing `LIST`.

dantleech committed:
Adding enhancers

dantleech commented:
Added the Enhancers at `Descriptor\Enhancer\{Sonata,Sylius}\{Admin,Resource}Enhancer` after all. Ultimately I think maybe having multiple test suites and modifying the `composer.json` file in Travis might be a solution if we have scaling issues.

dantleech committed:
Added Sylius Resource Enhancer

dantleech commented:
Added/implemented Sylius ResourceEnhancer.

dantleech commented on Description/DescriptionEnhancerInterface.php:
```diff
@@ -0,0 +1,35 @@
+    /**
+     * Return true if the provider supports the given type.
+     *
+     * @param CmfResource $resource
```
s/Cmf/Puli

dantleech commented on Description/Enhancer/Sonata/AdminEnhancer.php:
```diff
@@ -0,0 +1,120 @@
+ *
+ * @author Daniel Leech <daniel@dantleech.com>
+ */
+class AdminEnhancer implements DescriptionEnhancerInterface
```
Still need to "manually" test this or setup a full Sonata environment here for integration testing.

dantleech replied:
This may as well wait until this filters up to the `resource-rest-bundle`.

dantleech commented on Description/Enhancer/Sylius/ResourceEnhancer.php:
```diff
@@ -0,0 +1,116 @@
+        return true;
+    }
+
+    private function getRouteName(Metadata $metadata, $action)
```
Not used

dantleech commented on Tests/Unit/Description/DescriptionFactoryTest.php:
```diff
@@ -0,0 +1,92 @@
+                $description = $args[0];
+                $description->set('foobar', 'barfoo');
+            });
+        $this->enhanceer1->supports($this->resource->reveal())->willReturn(true);
```
spelling

dantleech commented on Tests/Unit/Description/DescriptionFactoryTest.php:
```diff
@@ -0,0 +1,92 @@
+    }
+
+    /**
+     * It should work when no enhancers are given.
```
s/give/provided

dantleech committed:
Fixes

dantleech commented:
@WouterJ could you have a look, I anticipate making further changes but I think it would be good to get this in to facilitate further work.

wouterj commented:
If everything works as expected from your point of view, let's just merge. We can improve it while it's in the master branch.

I agree with the concept behind this PR and don't see major problems in a quick review. I don't have time to test it atm.
